### PR TITLE
Add PySCENIC

### PR DESCRIPTION
**Pull request by LiliyaBioinf**
Describe your pull request here

----

Please read the [guidelines for Bioconda recipes](https://bioconda.github.io/contributor/guidelines.html) before opening a pull request (PR).

### General instructions

* If this PR adds or updates a recipe, use "Add" or "Update" appropriately as the first word in its title.
* New recipes not directly relevant to the biological sciences need to be submitted to the [conda-forge channel](https://conda-forge.org/docs/) instead of Bioconda.
* PRs require reviews prior to being merged. Once your PR is passing tests and ready to be merged, please issue the `@BiocondaBot please add label` command.
* Please post questions [on Gitter](https://gitter.im/bioconda/Lobby) or ping `@bioconda/core` in a comment.

### Instructions for avoiding API, ABI, and CLI breakage issues
Conda is able to record and lock (a.k.a. pin) dependency versions used at build time of other recipes.
This way, one can avoid that expectations of a downstream recipe with regards to API, ABI, or CLI are violated by later changes in the recipe.
If not already present in the meta.yaml, make sure to specify `run_exports` (see [here](https://bioconda.github.io/contributor/linting.html#missing-run-exports) for the rationale and comprehensive explanation).
Add a `run_exports` section like this:

```yaml
build:
  run_exports:
    - ...

```

with `...` being one of:

| Case                             | run_exports statement                                               |
| -------------------------------- | ------------------------------------------------------------------- |
| semantic versioning              | `{{ pin_subpackage("myrecipe", max_pin="x") }}`     |
| semantic versioning (0.x.x)      | `{{ pin_subpackage("myrecipe", max_pin="x.x") }}`   |
| known breakage in minor versions | `{{ pin_subpackage("myrecipe", max_pin="x.x") }}` (in such a case, please add a note that shortly mentions your evidence for that) |
| known breakage in patch versions | `{{ pin_subpackage("myrecipe", max_pin="x.x.x") }}` (in such a case, please add a note that shortly mentions your evidence for that) |
| calendar versioning              | `{{ pin_subpackage("myrecipe", max_pin=None) }}`    |

while replacing `"myrecipe"` with either `name` if a `name|lower` variable is defined in your recipe or with the lowercase name of the package in quotes.

### Bot commands for PR management

<details>
  <summary>Please use the following BiocondaBot commands:</summary>

Everyone has access to the following BiocondaBot commands, which can be given in a comment:

<table>
  <tr>
    <td><code>@BiocondaBot please update</code></td>
    <td>Merge the master branch into a PR.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please add label</code></td>
    <td>Add the <code>please review & merge</code> label.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please fetch artifacts</code></td>
    <td>Post links to CI-built packages/containers. <br />You can use this to test packages locally.</td>
  </tr>
</table>

Note that the <code>@BiocondaBot please merge</code> command is now depreciated. Please just squash and merge instead.

Also, the bot watches for comments from non-members that include `@bioconda/<team>` and will automatically re-post them to notify the addressed `<team>`.

</details>

## Summary by CodeRabbit

- **New Features**
	- Introduced a new package configuration file for `pyscenic`, detailing versioning and dependencies.
	- Added a build script to streamline the installation process for the `pyscenic` package.
- **Bug Fixes**
	- Updated import statements to utilize the standard `multiprocessing` library, improving compatibility and stability.

### CODE REVIEW

**Comment by pcm32** [comment]
This seems to be often failing (and then losing the logs) with:

```
11:52:46 BIOCONDA INFO (OUT) WARNING: No numpy version specified in conda_build_config.yaml.  Falling back to default numpy value of 1.22
11:52:46 BIOCONDA INFO (OUT) Adding in variants from internal_defaults
11:52:46 BIOCONDA INFO (OUT) Adding in variants from /opt/host-conda-bld/conda_build_config_0_-e_conda_build_config.yaml
11:52:46 BIOCONDA INFO (OUT) Adding in variants from /opt/host-conda-bld/conda_build_config_1_-e_bioconda_utils-conda_build_config.yaml
11:52:48 BIOCONDA INFO (OUT) Attempting to finalize metadata for pyscenic
##[warning]Free memory is lower than 5%; Currently used: 95.32%
##[warning]Free memory is lower than 5%; Currently used: 95.32%
##[warning]Free memory is lower than 5%; Currently used: 95.32%
##[warning]Free memory is lower than 5%; Currently used: 95.32%
##[warning]Free memory is lower than 5%; Currently used: 95.32%
##[warning]Free memory is lower than 5%; Currently used: 95.32%
```

on test linux...

**Comment by pcm32** [comment]
Getting past that and now sort of stuck in:

```
11:55:51 BIOCONDA INFO (OUT) Files containing CONDA_PREFIX
11:55:51 BIOCONDA INFO (OUT) -----------------------------
11:55:51 BIOCONDA INFO (OUT) bin/arboreto_with_multiprocessing.py (text): Patching
11:55:51 BIOCONDA INFO (OUT) bin/csv2loom (text): Patching
11:55:51 BIOCONDA INFO (OUT) bin/db2feather (text): Patching
11:55:51 BIOCONDA INFO (OUT) bin/gmt2regions (text): Patching
11:55:51 BIOCONDA INFO (OUT) bin/invertdb (text): Patching
11:55:51 BIOCONDA INFO (OUT) bin/pyscenic (text): Patching
11:55:58 BIOCONDA INFO (OUT) WARNING: Importing conda-verify failed.  Please be sure to test your packages.  conda install conda-verify to make this message go away.
11:56:02 BIOCONDA INFO (OUT) TEST START: /opt/conda/conda-bld/linux-64/pyscenic-0.12.1-py310hdbdd923_0.tar.bz2
11:56:04 BIOCONDA INFO (OUT) Renaming work directory '/opt/conda/conda-bld/pyscenic_1726055566916/work' to '/opt/conda/conda-bld/pyscenic_1726055566916/work_moved_pyscenic-0.12.1-py310hdbdd923_0_linux-64'
11:56:04 BIOCONDA INFO (OUT) shutil.move(work)=/opt/conda/conda-bld/pyscenic_1726055566916/work, dest=/opt/conda/conda-bld/pyscenic_1726055566916/work_moved_pyscenic-0.12.1-py310hdbdd923_0_linux-64)
```

**Comment by pcm32** [comment]
Stuck on the above now for 30 minutes... otherwise all looks good.

File: recipes/pyscenic/meta.yaml
Comment (pcm32) [inline recipes/pyscenic/meta.yaml:15]: @LiliyaBioinf arboreto is available from conda, and as such should be used as a conda dependency. Can you remove it from the sources and add under requirements.run please? Can you do the same with any of the above sources that might also be in conda (usually searching for conda + name of the package should allow you to find them if the exists as conda packages). Thanks!

Comment (pcm32) [reply recipes/pyscenic/meta.yaml:15]: https://anaconda.org/bioconda/arboreto

Comment (pcm32) [reply recipes/pyscenic/meta.yaml:15]: and of course, if added to the requirements, then shouldn't go in the `build.sh`.

**Comment by aliciaaevans** [comment]
We have had some resource issues, such as running out of memory, with Azure recently. After moving all dependencies out of source (either using existing bioconda or conda-forge packages or by adding new recipes in separate approved/merged PRs), if you are still seeing the Azure linux-64 build hang/disconnect, let me know and I will take a look.

File: recipes/pyscenic/meta.yaml
Comment (aliciaaevans) [reply recipes/pyscenic/meta.yaml:15]: Looks like there was some discussion a while back to add multiprocessing_on_dill to [conda-forge ](https://github.com/conda-forge/staged-recipes/pull/6199), but the upstream maintainers did not approve the [PR to add the license](https://github.com/sixty-north/multiprocessing_on_dill/issues/1). That would need to be resolved in some way so it can be added to conda-forge.

**Comment by pcm32** [comment]
@BiocondaBot please fetch artifacts

**Comment by pcm32** [comment]
Thanks @LiliyaBioinf for addressing the issues.

File: recipes/pyscenic/meta.yaml
Comment (pcm32) [reply recipes/pyscenic/meta.yaml]: maybe in the meantime multiprocessing_on_dill can be installed through pip so that the community can benefit soon from this recipe.

**Review by aliciaaevans — CHANGES_REQUESTED** [review]
Thanks for the recipe, but I don't think this should be packing its dependencies into the recipe as sources. One of the reasons is that license for distribution needs to be accounted for.

File: recipes/pyscenic/meta.yaml
Comment (pcm32) [reply recipes/pyscenic/meta.yaml:14]: agreed!

File: recipes/pyscenic/meta.yaml
Comment (LiliyaBioinf) [reply recipes/pyscenic/meta.yaml]: it was my github id, deleted it

File: recipes/pyscenic/meta.yaml
Comment (pcm32) [reply recipes/pyscenic/meta.yaml]: Working out a license like that could take a long time, in the meantime this could be used like this and possibly improved once that is resolved. There are many packages in bioconda that recourse to pip installing dependencies that are not available in conda, for instance https://github.com/bioconda/bioconda-recipes/blob/def3a07b68330abb40f7650b3747b8bee189ab66/recipes/latch/build.sh

**Comment by pcm32** [comment]
I suspect that you will need multiprocessing_on_dill from source, unless that someone has added to conda.

File: recipes/pyscenic/meta.yaml
Comment (aliciaaevans) [reply recipes/pyscenic/meta.yaml:13]: Thanks, I see your point. I will get some clarity on this from the core team.

**Comment by daler** [comment]
Chiming in here, to summarize it sounds like `mulitprocessing_on_dill` is not available on conda-forge because of a missing license in 2018, and an upstream author replied to a PR while on vacation in 2019 but probably didn't check afterwards.

The right way to fix this is to 1) submit a new conda-forge PR and 2) contact the upstream author(s) again. 

(Side note, the abovementioned [latch recipe](https://github.com/bioconda/bioconda-recipes/blob/def3a07b68330abb40f7650b3747b8bee189ab66/recipes/latch/build.sh) is the only instance I know of that is installing separately from pip, and is probably a fluke -- my guess is that it got missed in the review process -- and should be fixed.)

**Comment by pcm32** [comment]
License aside, I have to say I have seen many recipes doing the pip trick when there are no conda packages along the years, here is another one I found quickly https://github.com/bioconda/bioconda-recipes/blob/master/recipes/iphop/build.sh . 

I think it puts a large burden on contributors to having to create recipes recursively like this. What happens if multiprocessing_on_dill has 2 other deps that are not in conda? Are we asking the contributor to also have to deal with that? where do we stop?

**Comment by daler** [comment]
@pcm32 after discussing with the @bioconda/core team, the consensus is that pip should not be used to install additional packages from within recipes. The existence of existing recipes that do this is an oversight, so we will be fixing them. We will also update the docs to clarify this policy.

Regarding recursively adding recipes -- yes, that is exactly what we need to do. If `multiprocessing_on_dill` has more dependencies, then they need to be added to conda as well. In fact, in the early days of bioconda (and conda-forge) the majority of the work was in creating all of those recursive dependencies. It's a lot rarer nowadays, but sometimes it still does need to happen like in this case.

**Review by LiliyaBioinf — COMMENTED** [review]
changed multiprocessing

**Comment by aliciaaevans** [comment]
@LiliyaBioinf feel free to pause code rabbit by commenting `@ coderabbitai pause` (without a space after the `@`) if it's comments are annoying while you are making changes.

**Comment by pcm32** [comment]
@LiliyaBioinf I would suggest looking at other recipes in bioconda that patch code, to see how they deal with that. I would expect the patching to need to happen on a build.sh (but I might be wrong - so I suggest that you check at other recipes that do this).

File: recipes/pyscenic/build.sh
Comment (LiliyaBioinf) [reply recipes/pyscenic/build.sh]: @coderabbitai pause

File: recipes/pyscenic/meta.yaml
Comment (pcm32) [inline recipes/pyscenic/meta.yaml:14]: It seems to be that in other recipes they handle patches like this:

```
source:
  url: https://github.com/kcleal/gw/archive/refs/tags/v{{ version }}.tar.gz
  sha256: {{ sha256 }}
  patches:
    - fix-linker-issues.patch  # [linux]
    - fix-context.patch
```

and they don't seem to do anything on the build.sh. Can you please try like this and removing the patching part from the build (I suspect you might not need the build if this patching mechanism works).

**Comment by LiliyaBioinf** [comment]
@aliciaaevans can you help me, please, by review of  patching  multiprocessing_of_dill. I created a patch file like in saw in bioconda recepies, but after running more than 30 min it is usually canceled. What could be the problem? How do deal with that?

**Comment by aliciaaevans** [comment]
> @aliciaaevans can you help me, please, by reviewing of what I have done - patching multiprocessing_of_dill. I created a patch file like in saw in bioconda recepies, but after running more than 30 min it is usually canceled. What could be the problem? How do deal with that?

Looks like it might be running out of RAM (or possibly CPU, but between the number of dependencies and some other PRs with a similar error, I'm guessing RAM). The runners we use have `2 core CPU, 7 GB of RAM, and 14 GB of SSD disk space.` Have you tried building this locally to see if it completes?  I'll take a deeper look but at first glance it seems OK.

File: recipes/pyscenic/meta.yaml
Comment (aliciaaevans) [reply recipes/pyscenic/meta.yaml:64]: @coderabbitai Item 1, the multiprocessing_on_dill import is replaced by multiprocessing in the patch, `dill_patch.patch`. multiprocessing_on_dill should not need to be in the dependencies.

**Comment by LiliyaBioinf** [comment]
@aliciaaevans yes, i ran it locally and the package built. It  was on 1 core and 16 Gb RAM. It also seemed to me that it lacked memory here, because I tried different approaches of patching.

**Comment by pcm32** [comment]
Don't worry about the patching as long as it is correctly applied, memory wise the patching should be negligible.

**Comment by aliciaaevans** [comment]
We've switched from Azure to GitHub actions because of the memory issues. Hopefully it will complete now.

**Review by mencian — APPROVED** [review]

**Comment by mencian** [comment]
The recipe looks go to me, anything else to add @aliciaaevans?

**Review by aliciaaevans — APPROVED** [review]
Looks good. Thanks!

**Comment by aliciaaevans** [comment]
@BiocondaBot please fetch artifacts

**Pull request merged**
